substratevm/src/com.oracle.svm.jni/src/com/oracle/svm/jni/functions/JNIFunctions.java

```diff
@@ -640,9 +640,9 @@ static WordPointer GetDirectBufferAddress(JNIEnvironment env, JNIObjectHandle ha
 
     @CEntryPoint(exceptionHandler = JNIExceptionHandlerReturnMinusOne.class, include = CEntryPoint.NotIncludedAutomatically.class, publishAs = Publish.NotPublished)
     @CEntryPointOptions(prologue = JNIEnvEnterPrologue.class, prologueBailout = ReturnMinusOneLong.class)
-    static long GetDirectBufferCapacity(JNIEnvironment env, JNIObjectHandle hbuf) {
-        Buffer buffer = JNIObjectHandles.getObject(hbuf);
-        return buffer.capacity();
+    static long GetDirectBufferCapacity(JNIEnvironment env, JNIObjectHandle handle) {
+        Object obj = JNIObjectHandles.getObject(handle);
+        return (obj instanceof Target_java_nio_Buffer) ? ((Target_java_nio_Buffer) obj).capacity : -1;
```

Comment:
The JNI documentation says
> Returns -1 if the given object is not a **direct** java.nio.Buffer

So I think there is also a check missing whether this is a direct buffer or not. `GetDirectBufferAddress` has the same issue, it also needs some check for a direct buffer first.

Reply:
> So I think there is also a check missing whether this is a direct buffer or not.

The `instanceof` operation **IS** the check:
```java
return (obj instanceof Target_java_nio_Buffer) ? ((Target_java_nio_Buffer) obj).capacity : -1;
```

> `GetDirectBufferAddress` has the same issue, it also needs some check for a direct buffer first.

No it doesn't. It also [performs the check](https://github.com/oracle/graal/blob/b49fe19ffe7f4cbd60ae2a040d19871d89fbd67c/substratevm/src/com.oracle.svm.jni/src/com/oracle/svm/jni/functions/JNIFunctions.java#L630):
```java
if (obj instanceof Target_java_nio_Buffer) {`
```

Reply:
No, `Buffer` can both be a direct buffer or a heap-based buffer. So that check is not sufficient.

Reply:
Then how is one supposed to check for these two types?

Reply:
The best thing is usually to look at the implementation in HotSpot, and do the same kind of checks.

Reply:
HotSpot checks for interface `sun.nio.ch.DirectBuffer`. You will need to introduce a `@TargetClass` as alias.

Reply:
@peter-hofer why a `TargetClass`? `DirectBuffer` is `public` so can be used directly for `instanceof` checks.

Reply:
And since `DirectBuffer` is unrelated in the class hierarchy to `Buffer`, we actually need to do a `instanceof` check for both types. Otherwise a malicious implementation of `DirectBuffer` that does not extend `Buffer` could lead to a `ClassCastException` again.

Reply:
True, a `@TargetClass` is actually not needed and checking for both `Buffer` and `DirectBuffer` is the safest variant. @koutheir , please also add such a check for `GetDirectBufferAddress`.

Reply:
I might have done something incorrectly. I pulled changes and merged from the master branch, and applied the same changes again in order to resolve the conflicts. Can we continue in https://github.com/oracle/graal/pull/4838 ?